### DISCUSSION TITLE
Job backfill with explicit image from Dagster UI

### DISCUSSION THREAD

**Discussion by HynekBlaha** [dagster-io/dagster#19970] [Q&A]
**Why do I want to do it?** 
I believe it would help my team process backfills in parallel on rapidly changing staging environment.

### Details:
- Staging Dagster instance running in Kubernetes
- Team of 10 members deploying to a single code location
- Each team member wants to run backfill for a job he works on (image build from his branch)
- Each member would like to run backfill (queued) which would take long time (eg. 1 day) to process
- They would like to run multiple backfill jobs from various images in parallel

### 1. By default, job is executed with the image of current code location.
- Job runs with image that is current when it gets dequeued.
- ❌ On staging dagster, code location images change rapidly. This leads to executing job with different image than expected.
<img width="1414" alt="Screenshot 2024-02-22 at 12 25 55" src="https://github.com/dagster-io/dagster/assets/26007903/8e9cdc70-89df-4cd0-8ea8-89ac5c8e748a">

### 2. Hardcode image to job_tags
- The image tag is dynamically resolved in our CI.
- ❌ Hardcoding staging images is code is prone to errors (and not a good practise in general).
```
@job(executor_def=k8s_job_executor, tags={
    "dagster-k8s/config": {
        "container_config": {
            "image": "myregistry/myimage:mytag"
        }
    }
})
def my_k8s_job():
    my_op()
```

### 3. Set image from UI
- Each team member can explicitly set job image tag.
- Once submitted, another member can deploy different image to code location and submit another backfill with different image tag.
- In my opinion the best option.
- ❌ **However, when the job gets dequeued the image gets unexpectedly overridden with current code deployment image tag**.
<img width="703" alt="Screenshot 2024-02-22 at 12 33 18" src="https://github.com/dagster-io/dagster/assets/26007903/a0889778-3fa0-47a0-9a7a-6e6cb27d4ef2">
<img width="1290" alt="Screenshot 2024-02-22 at 12 35 59" src="https://github.com/dagster-io/dagster/assets/26007903/a4679b3d-754f-479e-a370-ef005fd86b96">

## Notes:
We tried other approaches, such as:
- Creating multiple code locations from the same codebase - it complains about same asset being defined multiple times.
- Splitting the codebase to multiple code locations - doesn't solve the problem, just reduces the probability of clash and increases the project complexity.

In my opinion, the solution 3) is a bug. If an image is provided as run tag from UI, it should be used.
Is there better approach that I overlooked?

Thank you for your hard work on Dagster. We love it.

**Comment by ion-elgreco** [comment]
We solved the issue by using asset prefix keys that has the name of our branch. This way we can deploy multiple code locations so that people can work in parallel on the same codebase and data.

**Reply by HynekBlaha** [reply]
Won't adding the branch name to asset prefix lead to creating many copies of the asset?
We often have historical data (eg. traffic records) materialized as upstream assets and we want to build upon them.
Starting with new (empty) version for each new branch isn't acceptable for us.
I believe allowing user to set the run image tag from UI is much straight-forward and less hacky solution.

**Reply by ion-elgreco** [reply]
@HynekBlaha that we solved using LakeFS, which will branch of from the main data branch for the feature. This way you can continue on top of the data.

I don't think the solution I proposed is hacky at all. It gives actually great seperation between assets that belong to a certain branch, runs and data

**Reply by HynekBlaha** [reply]
@ion-elgreco With the LakeFS it doesn't feel "hacky". Thank you for an interesting suggestion for workaround.

**Comment by aivanni** [comment]
I do not know why dagster team does not mention it anywhere, as this is a thing that can prevent adopting dagster in the first place.
It took me a lot of time to figure this out so sharing here. From Launchpad you can set `dagster-k8s/config` tag to a value e.g. 
`{"container_config": {"image": "<IMAGE NAME>:<IMAGE TAG>", "resources": {"requests": {"cpu": "4", "memory": "10Gi"}}}}`. Then the value for the image gets picked up by the launcher. 
And behavior for overwriting `dagster/image` tag has to be properly documented too. This thread is the first place I've seen that mentions `dagster/image` and not just `image` as it is shown in UI.

**Reply by HynekBlaha** [reply]
Been using this for a while now. Unfortunately doesn't get propagated to step jobs when using `executor_def=k8s_job_executor`.
